.Gruntfile.babel.js

```diff
@@ -179,7 +179,7 @@ module.exports = (grunt) => {
   grunt.registerTask('backend', ['backend:relaunch', 'watch'])
   grunt.registerTask('dev', ['checkDependencies', 'build:watch', 'connect', 'backend'])
   grunt.registerTask('dist', ['build'])
-  grunt.registerTask('test', ['dist', 'connect', 'exec:test'])
+  grunt.registerTask('test', ['dist', 'connect', 'exec:cypress', 'exec:test'])
```

Comment:
@greg I’m almost making it all work - `grunt test` runs both fe and be tests and the server only starts once.
But now the FE tests fail when running here. They start all fine it seems we can't do any logic. It throws several errors related to primus “error in connection establishment”. (?) When I try to signUp, on submit it fails with “SignUp.vue submit() error: TypeError: Failed to fetch”

To help you out:
- You can checkout the branch and run `npm test`.
- After the test fails it will generate screenshots and videos for you to analyze at `tests/cypress`.
- To make your life easier while debugging you can change the line 125:

```dif
- cypress: 'npm run cy:run',
+ cypress: 'npm run cy:open',
```

This will open the dashboard for you to see the tests running.

If you need to pause (debugger) at any time a test, just use `.pause()`. ex:

```diff
- cy.getByDT('signName').clear().type(userName)
// this will pause the test when trying to get signName so you have time
// to play around (you can even open other container tab in firefox
// to explore the app at the same time) and then continue the tests
+ cy.getByDT('signName').pause().clear().type(userName)
```

Reply:
So at least one issue here is that you're running `exec:cypress` as well as `exec:test`. Because of how Cypress is designed, we have to go all in with it, meaning we can't run `exec:test` which runs Mocha. Cypress is Mocha, plus Nightmare, plus a bunch of other stuff rolled into one.

So to make sure that the API server is started only once you might have to migrate the backend tests out of Mocha and into the Cypress system.

It's unclear whether we would get rid of Mocha completely or use Cypress to also test the `.test.js` files, since you'll note that's what we tell Mocha to do:

```
      test: {
        cmd: 'node node_modules/mocha/bin/mocha --require Gruntfile.js --exit -R spec --bail "{./{,!(node_modules)/**/}*.test.js,./test/*.js}"',
        options: { env: { LOAD_NO_FILE: 'true', ...process.env } }
      },
```

Basically, you need to pay attention to how we used to run tests (see above code) and make sure the same thing happens with Cypress, and that the minimum number of dependencies is run (removing Mocha if possible).

You'll note, for example, that when running tests we first make sure `Gruntfile.js` is loaded, and that for testing we set `LOAD_NO_FILE: 'true'` as part of the environment variables to prevent it from loading the `.Gruntfile.babel.js` file (or any other files). Why do we load this file? Because it's responsible for requiring `'@babel/register'` and setting certain environment variables that are used both on the frontend and backend.

You'll have to make sure it gets loaded first thing with Cypress as well.

Then you'll have to decide how to run Cypress. You could use `grunt-exec` like you're doing now, or you could define a grunt task [to programmatically run it](https://docs.cypress.io/guides/guides/module-api.html#cypress-run) directly. The latter option might be the better and more performant way to go, since the Gruntfile.js will have already been loaded, meaning all envars will have been defined and babel already registered (not needing to re-register it).

If figuring out how to get Cypress to run the `.test.js` files as well turns out to be too difficult, then you can create an issue to do that later, and run Mocha on just those files (again: the backend tests still would need to be migrated to Cypress to make sure the API server is loaded only once per `grunt test` invocation).

Reply:
Nice, there is a [`testFiles` config variable](https://docs.cypress.io/guides/references/configuration.html#Global) you can pass in via [`cypress.run()`](https://docs.cypress.io/guides/guides/module-api.html#cypress-run), so you might be able to get rid of mocha and run `*.test.js` files via Cypress as well, and maybe even get rid of its `cypress` directory structure too which we don't need.

Reply:
Another thought: it might be worth calling `require('cypress')` *within the grunt task* instead of at the top of the file like usual, as loading Cypress each time `grunt` is called might slow down the grunt command unnecessarily for other tasks like `grunt dev`, `grunt clean`, etc.

Reply:
I appreciate your effort in trying to explain this to me but I still don't understand how grunt/backend stuff works.

I spent the last hour around trying to tweak this and FE never passed with a proper solution for the same reason: it can't connect to the backend, even after requiring Gruntfile.js with `LOAD_NO_FILE` as true.

It sucks tht my experience with this config stuff is not as good as with the rest of frontend :/

Reply:
Okay, I gotcha what's wrong, it was not the gruntfile part. It was the test responsible to start the server:

```
  it('Should start the server', function () {
    this.timeout(10000)
    return require('../backend/index.js')
  })
```

This isn't even a test... this is a hacky way of starting the server so the next tests run fine.
As far as I understand, this should be done before starting cypress or mocha.
And that's what I'm trying to do now, but because it's written in ES (with import), it first needs to be transpiled. Working on it now... I hope when you wake this is all fixed

Reply:
Well, I think I'm almost there (at least the tests are passing with just one server running!).

So basically I added `backend:relaunch` task to test task so it launches the server, but it seems it does more than the necessary (not sure, you tell me!)...

The only issue now is that when running it on my pc the tests end like this:

```bash
...
    ✓ don't distribute anything if everyone is above mincome


  66 passing (204ms)


Done.
message received in child, shutting down...
➜  group-income-simple git:(add-cypress) Hapi server down
[server] Unhandled exception: Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at process.target.send (internal/child_process.js:636:16)
    at Primus.send (/Users/sandrina-p/Documents/git/group-income-simple/backend/index.js:44:17)
    at process._tickCallback (internal/process/next_tick.js:68:7) Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at process.target.send (internal/child_process.js:636:16)
    at Primus.send (/Users/sandrina-p/Documents/git/group-income-simple/backend/index.js:44:17)
    at process._tickCallback (internal/process/next_tick.js:68:7)

```

So it seems something is still running and it doesn't shut down. (I need to do it manually ctrl+C). I couldn't understand exactly how to fix that. But it doesn't happen on Travis.

Reply:
> It's unclear whether we would get rid of Mocha completely and use Cypress to also test the .test.js files.
> (...)
> that the minimum number of dependencies is run (removing Mocha if possible).
> (...)
> If figuring out how to get Cypress to run the .test.js files as well turns out to be too difficult, then you can create an issue to do that later, and run Mocha on just those files.

Migrating our backend tests to Cypress doesn't mean we would get rid of Mocha, because Cypress is built on top of Mocha, so we would still have that dependency on our project.

Also, I think Cypress is not meant to run functional unit-tests, only end-to-end tests. (See this for more info: https://github.com/cypress-io/cypress/issues/318). It's like saying "Let's use Nightmare to run backend tests too and get rid of Mocha", it doesn't work like that... 

For that reason, I see no problem in keeping the unit-tests being run by Mocha only without enforcing extra work.

Reply:
> This isn't even a test... this is a hacky way of starting the server so the next tests run fine.

Believe it or not it is a test, and IMO it's not hacky at all 😄. If you look into `backend/index.js` you'll see:

```js
module.exports = new Promise((resolve, reject) => {
  sbp('okTurtles.events/on', SERVER_RUNNING, function () {
    console.log(chalk.bold('backend startup sequence complete.'))
    resolve()
  })
})
```

This is how the test is run. It will fail by timeout if the server fails to start for some reason (which is possible), and it will pass if the server started correctly. It should also continually pass from that point on, which is why both the backend and the frontend tests used to have it. This should be re-introduced here, so that the backend server is started within Cypress, and not by Grunt. The reason you are running into the ^C problem is because you're using `backend:relaunch` to start the backend server. That task is only for `grunt dev`, instead the backend server should be started as before, within the tests, and both by the frontend and the backend tests (but only once overall, which is automatically accomplished as I explained in a previous comment about the test that returns `return require('../backend/index.js')`.

Furthermore, you need to group the two tests (backend and frontend) together in the same process. By using `exec:test` and `cypress` tasks, you're running them in two separate processes. This unnecessarily slows down the tests, and also keeps the `grunt test` command running (because of `backend:relanch`). Let's see what it's like to take advantage of the fact that Cypress has Mocha and run the backend tests inside of Cypress in the same process. You will be able to remove `backend:relaunch` then from the tasks.

You can then try and see if there's a way to configure the `cypress.run()` command to also run the `.test.js` unit tests that we have in the project. It might be that it's impossible because Cypress is forced to search only within the `test/cypress` directory, but maybe you can discover a way to get it to search across the entire project per the `testFiles` config option I mentioned and linked to above. If it's possible to do this, it means we can get rid of `exec:test` completely, and have only a single nodejs process launched on `grunt test`.

Finally, don't forget to make sure that your `cypress` task checks either [`this.flags`](https://gruntjs.com/api/inside-tasks#this.flags) or [`grunt.option`](https://gruntjs.com/api/grunt.option) to run either `cypress.run()` or `cypress.open()`, so that we can use `grunt test` to run `cypress.run()` and `grunt test:browser` or `grunt test --browser` to display them live.

Reply:
If we end up having `.test.js` tests run by `exec:test`, that is OK, but I would like to at least see if it's possible to run them within Cypress because then we can compare the speed of the two approaches. The link you gave me claims that it will be slower to run the unit tests in Cypress, but I'd like to see how much slower. It might be that it's not much of a slowdown compared to launching a second process via `grunt-exec`.

Reply:
> I would like to at least see if it's possible to run them within Cypress

To be clear: I mean run them within Cypress but leaving those files exactly where they are in the project (not moving them into the cypress/ folder).

Reply:
> This is how the test is run. It will fail by timeout if the server fails to start for some reason (which is possible), and it will pass if the server started correctly. It should also continually pass from that point on, which is why both the backend and the frontend tests used to have it. 

Okay, now I understand why we need the test... Still, we shouldn't have to write it twice, it's a unit test, so doing it only once is the correct.

Reply:
Okay, so my solution at the moment:

- Readded test to the backend.js to the unit tests (`test/backend.js`) as it was before.
- In the end-to-end tests added two tasks responsible to require gruntFile and to start the server:

```
  grunt.registerTask('test:e2e', ['dist', 'connect', 'backend:launch', 'exec:requireGruntfile', 'cypress'])
```

The issue now is in one particular test (update profileDisplayName) that fails. The profile settings are saved but the name isn't updated on the sidebar unless I refresh the page... In the console I have this error:

```
primus.js:3572 WebSocket connection to 'ws://localhost:3000/primus?_primuscb=Mo-BBJH' failed: Error during WebSocket handshake: Unexpected response code: 404
```

Reply:
I even tried to do like you said: adding the backend test to cypress as well to start the server, but that file (and its imports) have FlowJS included and Cypress can't run those files unless we add some preprocessor to it, which I couldn't find a way of doing easily :/

Reply:
The good news is I've managed to show browser while running the tests, bu using grunt flags:

```bash
# tests run on "run" mode by default 
grunt test:e2e

# run tests headed (shows the browser).
grunt test:e2e:headed

# run tests on open mode - which is different from headed
# open allows you to interact, debug, choose what tests to run
# headed only shows you the browser.
grunt test:e2e:open
```

Reply:
Another thing that I don't understand yet is:
In the backend tests, we start (and test) the server by requiring `backend.js` and waiting for its promise to be done. But how does it know when to stop the server? (which happens after the unit tests finish).